src/display_context.cpp

```diff
@@ -136,32 +137,43 @@ int display_context::village_owner(const map_location& loc) const
  */
 bool display_context::is_observer() const
 {
+#ifdef __cpp_lib_ranges
+	return std::ranges::none_of(teams(), &team::is_local);
+#else
```

Comment:
imo there is no benefit of adding conditionally compiled stuff like this now indead over doing it when all complilters we use support it, while on the other hand it does increase maintance work in case someone wants to change this code.

Also in the partiuclar case you can probably just use boost::algorithm::none_of https://www.boost.org/doc/libs/1_86_0/libs/algorithm/doc/html/algorithm/CXX11.html#the_boost_algorithm_library.CXX11.CXX11_inner_algorithms.none_of and have te same code withotu the need for having multiple codes that do the same thing in our codebase.

Reply:
I'm not keeping the ranges stuff in this PR, at least. It was a local experiment. Need to decide how to handle it more broadly. The boost version wouldn't be a drop-in replacement for all cases, though, since it doesn't accept a projection argument.

Reply:
What do you mean by a "projection argument"?

Reply:
Just learned about them recently. Apparently almost all the new constrained algorithms take a "projection" argument besides the predicate function which transforms the view before you examine it through the predicate. It means any use of `func_if` where you provide a predicate lambda in order to examine the value of a specific member or return of a specific member function can be replaced with just `func`.

So, for instance, code such as this
```cpp
// recall_unit_manager
auto it = std::find_if(recall_list_.begin(), recall_list_.end(), [&unit_id](const unit_ptr& ptr) { return ptr->id() == unit_id; });

// display_context::side_units
return std::count_if(units().begin(), units().end(), [side](const unit& u){ return u.side() == side; });
```
becomes this:
```cpp
auto it = std::ranges::find(recall_list_, unit_id, &unit::id);

return std::ranges::count(units(), side, &unit::side);
```

Reply:
This makes me think of Python, where you provide a "key function" rather than a comparator for sorting. Obviously it's not quite the same, as this is searching, not sorting, but it seems like a similar concept.

Reply:
`std::ranges::sort` also accepts a projection argument. The notable exception I know of is `fold_left`.